### DISCUSSION TITLE
How to use PTB Diagnostic ECG Database on Neurokit 2

### DISCUSSION THREAD

**Discussion by IHuaT** [neuropsychology/NeuroKit#508] [Q&A] [answered]
Hi

I am working on the  PTB Diagnostic ECG Database for feature extraction. I try to extract 15 interval features and 18 amplitude features. The part I am stuck with right now is that I trying to use PTB Diagnostic ECG Database within Neurokit2 but the program shows the HTTPError. Do you have any idea for implement the PTB dataset? 
Here are two links I use for PTB dataset download.
https://www.physionet.org/content/ptbdb/1.0.0/
https://archive.physionet.org/cgi-bin/atm/ATM

**Comment by zen-juen** [comment, marked as answer]
Hi @as1123000 how does your script for downloading the PTB database with neurokit look like?
In the meantime, perhaps you can try out this short script for the ptbdb 1.0.0 database

```

import pandas as pd
import numpy as np
import wfdb
import os

# set working directory first
wd = os.getcwd()

patient_files = [file for file in os.listdir("ptb-diagnostic-ecg-database-1.0.0/") if 'patient' in file]

data_files = []
for participant in patient_files:
    path = wd + "\\ptb-diagnostic-ecg-database-1.0.0\\" + participant
    filename = [i for i in os.listdir(path) if ".dat" in i][0]
    file = path + "\\" + filename

    # get signal
    data = pd.DataFrame()
    for i in range(15):
        signal_name = wfdb.rdsamp(record_name=file[:-4])[1]['sig_name'][i]
        signal = wfdb.rdsamp(record_name=file[:-4])[0][:, i]
        channel = pd.DataFrame({str(signal_name): wfdb.rdsamp(record_name=file[:-4])[0][:, 0]})
        data = pd.concat([data, channel], axis=1)

    data["Participant"] = participant
    data["Sample"] = range(len(data))
    data["Sampling_Rate"] = 1000
    data["Database"] = "PTB-Diagnostic-ECG-Database-1.0.0"
    data["Sex"] = wfdb.rdsamp(file[:-4])[1]['comments'][1]
    data["Age"] = wfdb.rdsamp(file[:-4])[1]['comments'][0]
    # can add more descriptive columns from the data 
    
    # append
    data_files.append(data)

# Tidy final output
df = pd.concat(data_files)
```

**Reply by IHuaT** [reply]
Hi

Thank you for the quick reply. 
The script I use is https://physionet.org/content/ptb-xl/1.0.1/example_physionet.py
It is provided from the  data_ptb_xl_download_ptbxl.py
I do not quite understand how to make the script working on the Neurokit program. I have followed the file's description of accessing the files. 
Would you mind share your idea with me?

I really appreciate your help.

**Reply by zen-juen** [reply]
Hi @as1123000 just to confirm with you, the `example_physionet.py` script seems like its specifically targetted for the PTB-XL database (https://physionet.org/content/ptb-xl/1.0.1/) rather than PTB 1.0.0 (https://www.physionet.org/content/ptbdb/1.0.0/). Is this PTB-XL database specifically what you want? 

If yes, we actually meant to write a [script](https://github.com/neuropsychology/NeuroKit/blob/master/data/ptb_xl/download_ptbxl.py) for this @DominiqueMakowski so we'll continue working on this!

**Reply by IHuaT** [reply]
Hi

The PTB-XL database is not actually what I want. I more prefer to use ptbdb 1.0.0 database.  I put the script of ptbdb 1.0.0 database on the Neurokit program, but I am not sure how to make the script working.

**Reply by zen-juen** [reply]
Thanks for the clarification! I think `example_physionet.py` is designed specifically for the 1.0.1 database and you probably can't directly use this on the 1.0.0 database because the structure of these two databases are different.

For starters, you can try running the short script that I shared above to access the patient files, of which the final output is a dataframe of signals concatenated across the different patients. You'll just need to set the working directory of your python script to the same location as where your PTB database and then run the script - let us know if this works!

**Reply by IHuaT** [reply]
Thank you so much for the patient and detailed description. I will keep working on it, and let you know any updates. Again, really appreciate your help.

**Comment by zen-juen** [comment]
Also, realized this database is not accounted for in https://github.com/neuropsychology/NeuroKit/tree/master/data do we want to add a script for downloading/processing it @DominiqueMakowski ?

**Reply by DominiqueMakowski** [reply]
I think so yes

**Comment by Salar-Sali** [comment]
@zen-juen @as1123000 
Good evening everybody!
I am trying to classify  PTB Diagnostic ECG Database Version: 1.0.0 but there are 8 classes and a normal class in this dataset.
now I am stuck at the beginning of the process, I don't know how to **LABEL** the data because for each patient I am reading the .hea file to know the diagnosis. is there another way to **label** the data which can save time?
